### ISSUE TITLE
Plugin `afterUpdateSpecificCoupons` isn't working as it should be. Wrong coupons are updated.

### ISSUE THREAD

**Issue by jasperzeinstra** [dotmailer/dotmailer-magento2-extension#526]
There's a big bug in de the plugin `\Dotdigitalgroup\Email\Plugin\CouponPlugin::afterUpdateSpecificCoupons`. It saves the `$subject` by using the values of the `$rule`. `$subject` is the coupon record, `$rule` is the sales rule record.

```
    /**
     * @param \Magento\SalesRule\Model\ResourceModel\Coupon $subject
     * @param \Magento\SalesRule\Model\ResourceModel\Coupon $result
     * @param \Magento\SalesRule\Model\Rule $rule
     *
     * @return \Magento\SalesRule\Model\Rule
     */
    public function afterUpdateSpecificCoupons(
        \Magento\SalesRule\Model\ResourceModel\Coupon $subject,
        $result,
        \Magento\SalesRule\Model\Rule $rule = null
    ) {
        if ($rule) {
            //update the generated and the expiration date
            $rule->setData('expiration_date', null);
            $rule->setData('generated_by_dotmailer', '1');
            $subject->save($rule);
        }

        return $rule;
    }
```

*How to reproduce*
* Set up a Magento 2.2.6 with sample code
* By default DotMailer isn't enabled, keep it like that
* Check the table `salesrule_coupon` in the database
* Go to sales rule and open `ID 1` ("Buy 3 tee shirts and get the 4th free")
* Set coupon type to `Specific coupon` and check `Use Auto Generation`
* Save sales rule and check database table `salesrule_coupon`
* Notice that the coupon with `coupon_id = 1` has changed, but this coupon is connected to `rule_id = 4`. But we saved `rule_id = 1`.
* Error is caused in `\Dotdigitalgroup\Email\Plugin\CouponPlugin::afterUpdateSpecificCoupons` where data of rule object is used to save coupon object.
* Also if DotMailer isn't active I wouldn't expect the variable `generated_by_dotmailer` to be set to `1`.

![dotmailer-db-1](https://user-images.githubusercontent.com/9214557/47563107-fbb62800-d920-11e8-8e61-861233ef8829.png)

![dotmailer-rules](https://user-images.githubusercontent.com/9214557/47563125-040e6300-d921-11e8-8308-b1f2672e4f84.png)

![dotmailer-rule-save](https://user-images.githubusercontent.com/9214557/47563127-07a1ea00-d921-11e8-8ffc-a2465ad56af6.png)

![dotmailer-db-2](https://user-images.githubusercontent.com/9214557/47563116-007adc00-d921-11e8-8a94-167ff3642f1a.png)

From my point of view you should check if DotMailer is enabled. There's a helper being defined in the constructer method, but it's not being used in the class. And you should not change the rule, but only the coupon.

If you take a look at how they save date for the coupon in `\Magento\SalesRule\Model\ResourceModel\Coupon::updateSpecificCoupons` you'll notice that they use the `update` method. So I think it's also better to use this code in the plugin of DotMailer

```
            $this->getConnection()->update(
                $this->getTable('salesrule_coupon'),
                $updateArray,
                ['rule_id = ?' => $rule->getId()]
            );
```

**Comment by simon-letch** [comment]
Hey @jasperzeinstra,

We're taking a look at this during our current sprint. I'll update again once we've investigated. From a quick initial glance, I can't see why this plugin is required at all.

Simon

**Comment by simon-letch** [comment]
This has now been released in v3.0.3.
